File: mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/memory/capture/CaptureFilter.kt
/** Filters for auto-capture — determines what should be captured. */
package com.codeintel.memory.capture

/** Determines if content is worth capturing based on heuristics. */
object CaptureFilter {

    private val DECISION_KEYWORDS = listOf(
        "decided", "chose", "selected", "approach", "architecture",
        "design decision", "trade-off", "alternative"
    )

    private val ERROR_KEYWORDS = listOf(
        "error", "failed", "exception", "bug", "fix", "root cause",
        "workaround", "solution", "resolved"
    )

    /** Check if text contains decision-related content. */
    fun isDecisionContent(text: String): Boolean {
        val lower = text.lowercase()
        return DECISION_KEYWORDS.any { lower.contains(it) }
    }

    /** Check if text contains error pattern content. */
    fun isErrorContent(text: String): Boolean {
        val lower = text.lowercase()
        return ERROR_KEYWORDS.any { lower.contains(it) }
    }

    /** Check if content is substantial enough to capture. */
    fun isSubstantial(text: String, minLength: Int = 50): Boolean {
        return text.trim().length >= minLength
    }

    /** Determine the best knowledge type for content. */
    fun classifyContent(text: String): String {
        return when {
            isDecisionContent(text) -> "DECISION"
            isErrorContent(text) -> "ERROR_PATTERN"
            text.contains("```") -> "PROCEDURE"
            else -> "CONTEXT"
        }
    }
}

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/memory/capture_filter.py
"""CaptureFilter — determines what content is worth auto-capturing."""

from __future__ import annotations

_DECISION_KEYWORDS = [
    "decided", "chose", "selected", "approach", "architecture",
    "design decision", "trade-off", "alternative",
]

_ERROR_KEYWORDS = [
    "error", "failed", "exception", "bug", "fix", "root cause",
    "workaround", "solution", "resolved",
]


def is_decision_content(text: str) -> bool:
    """Check if text contains decision-related content."""
    lower = text.lower()
    return any(kw in lower for kw in _DECISION_KEYWORDS)


def is_error_content(text: str) -> bool:
    """Check if text contains error pattern content."""
    lower = text.lower()
    return any(kw in lower for kw in _ERROR_KEYWORDS)


def is_substantial(text: str, min_length: int = 50) -> bool:
    """Check if content is substantial enough to capture."""
    return len(text.strip()) >= min_length


def classify_content(text: str) -> str:
    """Determine the best knowledge type for content."""
    if is_decision_content(text):
        return "DECISION"
    if is_error_content(text):
        return "ERROR_PATTERN"
    if "```" in text:
        return "PROCEDURE"
    return "CONTEXT"
